WIP Draft: Shimcache execution timeline feature with Amcache enrichment

Markus98 opened the pull request:
Quite a lot of new stuff compared to last time. Implemented the initial timelining feature for windows 10 versions of shimcache and amcache. This code is not ready for a real PR, but it would be valuable to get some feedback at this point.

Some things that would be good to get input on:
- Should I split the source files that I have been working on further? For example the `file/hve.rs` file might get quite bloated if parsing for all of the different shimcache and amcache versions are put there.
- They way I am handling outputting to csv at the moment is probably not perfectly in line with the rest of the program, so any pointers on how that should be done in a proper way
- Any generic pointers regarding Rust common practices that I might be breaking
- Are the params for the timeline command appropriate and in line with the rest of the program?
- If there are any issues with naming, clarity, or code comments
- Does the error handling look good enough or should it be more precise?

Stuff I will clean up anyways:
- Will most likely remove unit tests unless I write proper ones
- Any rogue comments or `println!`s that I have not removed yet

alexkornitzer reviewed (commented):
On the whole seems pretty good (obviously I have not tried it with samples). My biggest concern is that the timeline functionality is not generic but maybe it never can be.

To address your questions above:
- Rust files can get quite big and that is fine, if you look at other crates you will see this. I think at the end of the day it comes down to readability and how well the code works in isolation etc.
- Yep you should be using the csv lib that is used in `cli`.
- So far your Rust seems very Rust apart from some bits I commented on, a good way is to run `cargo clippy` and to make sure that you have the rust formatter running. The rest will come with time and reviews.
- I think that params are fine for hve but not for anything else.
- Most naming is good, I would say cause I don't know the hve format that following the parser is hard and could benefit with comments.
- Error handling looks sufficient for what we need we can just expand it as required if that is ever the case.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+    pub programs: HashMap<String, ProgramArtifact>,
+}
+
+pub struct AmcacheFileIterator<'a> {
```
It's because of the way I structured the entities of Amcache. There are two entry types in amcache: programs and files. Programs have a unique ID, and the file entries have a reference to a program which they belong to.

So amcache has a data structure which is something like this:
```js
{
  "example_program_id": {
    "program_entry": { ... },
    "file_entries": [ ... ]
  },
  ...
}
```

So the `program_iterator` is for iterating the outer hashmap elements and the file_iterator is for the inner `"file_entries"` vector of each program.

So it's not exactly n^2, but the logic behaves the same. In reality `n = number of file entries in amcache`.

Markus98 commented on Cargo.toml:
```diff
@@ -32,6 +32,7 @@ serde_yaml = "0.9"
 tau-engine = { version = "1.0", features = ["core", "json"] }
 term_size = "0.3"
 uuid = { version = "1.1", features = ["serde", "v4"] }
+notatin = { git = "https://github.com/strozfriedberg/notatin", rev = "9783169f4649c8df92f94b8c20421df8f4aa070a" }
```
Moved the import to the proper alphabetical order.

Yeah I know, it's not ideal using an unreleased crate, but the lib saved tons of time. Also avoided having to wrap non-rust applications. Hopefully we can do something about this in the future.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+            Some(file) => Some(file),
+            None => {
+                loop {
+                    let next_program = self.program_iterator.next();
```
Reduced 6 lines of code and made it more readable, nice.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+    type Item = &'a InventoryApplicationFileArtifact;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next_file = match &mut self.file_iterator {
```
Seems like it, neat. Changed.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+            })
+    }
+
+    //TODO: parsing of different versions of Amcache
```
Will do and fixed existing comments.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+                None
+            };
+
+            let mut program_artifact = ProgramArtifact::new(&program_id);
```
Yeah, there are cases when the `ProgramArtifact` has no app_artifact. This happens when there is an entry of a file that has a reference to a `program_id` which was not in the list of program entries. This case is handled later on in the `parse_amcache` function and creates a new `ProgramArtifact` which has no app artifact.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+                None
+            };
+
+            let sha1_hash = String::from(&file_id[4..]);
```
Added some safety measures and comments

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+            let cache_signature = std::str::from_utf8(&shimcache_bytes[offset_to_records..offset_to_records+4])?;
+            // Windows 10 shimcache
+            if cache_signature == "10ts" {
+                let re = Regex::new(r"^([0-9a-f]{8})\s+([0-9a-f]{16})\s+([0-9a-f]{16})\s+([\w]{4})\s+([\w.]+)\s+(\w+)\s*(\w*)$").unwrap();
```
Usually called only once, but made it `lazy_static` anyways for good measure. Gotta leave the unwrap in in that case though.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+    }
+}
+
+fn utf16_to_string(bytes: &[u8]) -> Result<String> {
```
Yes

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+    Ok(NaiveDateTime::parse_from_str(ts_str, "%m/%d/%Y %H:%M:%S")?)
+}
+
+#[cfg(test)]
```
Yeah there were only for initial testing before I had an implemented the features to the cli. I was going to remove them before the actual PR.

Also regarding test data, should that be provided in the repository itself or something that is provided externally?

alexkornitzer replied:
Internally, but basically the `tests` folder is only for testing stuff that is publicly exposed (in the `lib.rs` file). The in file tests are there to test the more unit like code.

Markus98 commented on src/main.rs:
```diff
@@ -203,6 +203,22 @@ enum Command {
     },
+
+    /// Create an execution timeline by combining entries from shimcache and amcache
+    Timeline {
```
Yeah let's think of a better name and description + any other considerations

alexkornitzer replied:
We discussed this offline, the plan just for public visibility is to move this to `analyse shimcache`.

Markus98 commented on src/main.rs:
```diff
@@ -203,6 +203,22 @@ enum Command {
+        /// A path to output results to.
+        #[arg(short = 'o', long = "output")]
+        output: Option<PathBuf>,
+    },
```
Duh, good catch.

Markus98 commented on src/timeline.rs:
```diff
@@ -0,0 +1,253 @@
+#[derive(Debug, Clone)]
+enum TimelineTimestamp {
+    Exact(DateTime<Utc>),
+    Range(DateTime<Utc>, DateTime<Utc>), // (from, to)
```
Maybe that is a good idea.

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+        let mut shimcache_entries: Vec<ShimCacheEntry> = Vec::new();
+
+        
+        let sig_num = u32::from_le_bytes(shimcache_bytes[0..4].try_into()?);
```
Added

Markus98 commented on src/file/hve.rs:
```diff
@@ -0,0 +1,421 @@
+        let shimcache_key = self
+            .inner
+            //TODO: get control set dynamically instead of hardcoded
+            .get_key("ControlSet001\\Control\\Session Manager\\AppCompatCache", false)?.unwrap();
```
This is going to be fixed along with the TODO above.

Markus98 commented:
Some more small changes:

### Changes since review 1
- Changed "timeline" command to "analyse shimcache"
  - Moved `timeline.rs` to `analyse/shimcache.rs`
- Moved csv ouput function to `cli`
  - Use `prettytable` library instead

alexkornitzer commented on src/main.rs:
```diff
@@ -203,6 +203,30 @@ enum Command {
+        shimcache: PathBuf,
+        /// The path to the newline delimited file containing regex patterns to match
+        #[arg(short = 'r', long = "regexfile")]
+        regex_file: PathBuf,
```
I think we should allow `-e|--regex` like above to feed multiple regex from the command line, and allow a file of regex as well. This removes the need to make a `config` file all the time.

alexkornitzer commented on src/file/hve/shimcache.rs:
```diff
@@ -0,0 +1,508 @@
+        };
+
+        // Windows XP shimcache
+        if signature_number == 0xdeadbeef {
```
We should probs break all these ifs out but its nothing major tbh.

Markus98 replied:
Do you mean into separate files? I was thinking about doing that as well.

alexkornitzer replied:
could be separate files, or just separate functions. you could even just `mod win xp { fn parse...}` in the one file, maybe that is the nicest way?

alexkornitzer replied:
You can write this as a match, and then use `_` to catch the unimplemented paths.

alexkornitzer reviewed (changes requested):
On the whole seems good to go, its hard to comment on some bits just to the sheer amount of parsing going on. I think you need to run clippy and rustfmt on this stuff too.

Markus98 commented on src/analyse/shimcache.rs:
```diff
@@ -0,0 +1,272 @@
+        }
+    
+        /// Sets timestamp ranges for timeline entities based on shimcache entry order and indices
+        fn set_timestamp_ranges(range_indices: &Vec<usize>, timeline_entities: &mut Vec<TimelineEntity>) {
```
Good point. Changed it so that no work gets done if `range_indices` is empty.

alexkornitzer replied:
What about the panic we can get at the end, should be checked too?

Markus98 replied:
If you are referring to the `.last().expect("...")`, then that should not have to be checked. The function immediately returns from the function at the start now if the `range_indices` length is zero. Therefore there should always be at least one element when calling `.last()` on `range_indices`.

alexkornitzer reviewed (changes requested):
Minor minor stuff

Markus98 commented on src/analyse/shimcache.rs:
```diff
@@ -0,0 +1,272 @@
+        fn extract_ts_from_entity(entity: &TimelineEntity) -> DateTime<Utc> {
+            match &entity.timestamp {
+                Some(TimelineTimestamp::Exact(timestamp, _type)) => *timestamp,
+                _ => panic!("Provided entities should only have exact timestamps!"),
```
From examples of `unreachable!` it seemed that it was only used in a context where it is truly unreachable within the function as well. The logic that makes the condition unreachable in this case is outside the function, so maybe using `unreachable!` is not completely appropriate here?

alexkornitzer reviewed (changes requested):
Some more minor things.

alexkornitzer commented:
Closing this out as we will cover in a new PR.
